### **Prediction** — a block-group crime model that beats the incumbent *inside* cities

The deployed **agency-scale** model (`04_bg_comparison`) levels whole cities well but fits weakly
*within* a city — it can't tell a city's dangerous blocks from its safe ones. This notebook builds a
block-group model on **`log(weighted rate)`** (absolute, no demeaning) and shows it closes that gap.

**Order of the story:**
1. **The incumbent's block-group problem** — weak block-group fit, for *both* property and total crime
   (Detroit is the sharpest case), taken as the reference row from `04`.
2. **The refresh on PROPERTY crime (9 cities)** first — the wider sample makes every number less
   brittle and shows *sample size is the main lever*.
3. **The same refresh on TOTAL crime (5 cities).**
4. **Per-city breakdown, hyperparameter tuning, and next steps.**

Every configuration is scored on two axes, the same ones `04` used to indict the incumbent:

| axis | question | metrics |
| --- | --- | --- |
| **CALIBRATION** | is each block group placed at the right crime rate? | `r2_oos`, `mae` |
| **RANK** (within-city) | inside a city, are the dangerous blocks flagged? | `recall@25`, `called_safe` |

- **Splits:** stratified **80/20** (interpolation — a city we've already labeled) and **LOCO**
  (leave-one-city-out — a brand-new city, the honest national-rollout estimate). The gap between them
  is the value of having seen a city before.
- **Predictor ladder:** `ours` → `ours+approved` (de-duplicated incumbent features) →
  `ours+approved+agency` (adds last year's agency-wide crime total). No race / protected predictors.
- **Models:** `RidgeCV` and `LightGBM`, same ladder.


In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

import contextlib
import io
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import optuna

_here = Path.cwd().resolve()
_root = next(p for p in [_here, *_here.parents] if (p / "src").is_dir())
sys.path.insert(0, str(_root / "src"))

from crime_blockgroup_mapping.constants import CITIES
from regression_modelling.constants import AGENCY_ANCHOR_COL
from regression_modelling.models import baseline
from regression_modelling.models.cv import load_pooled_table, loco_metrics, within_city_recall
from regression_modelling.models.experiments import (
    ExperimentSpec, run_spec, resolve_predictors,
)

pd.set_option("display.width", 240)
pd.set_option("display.max_columns", 40)
warnings.filterwarnings("ignore", category=RuntimeWarning)
warnings.filterwarnings("ignore", message="An input array is constant")
optuna.logging.set_verbosity(optuna.logging.WARNING)

In [2]:
# 5 full-coverage cities carry weighted-total (wtotal); the rest add weighted-property (wprop).
full_cities = [c for c, cfg in CITIES.items() if not cfg.property_only]
all_cities = list(CITIES.keys())

# Pooled block-group tables: one row per BG with predictors, observed rates, and the agency total.
pool5 = load_pooled_table(full_cities)
pool10 = load_pooled_table(all_cities)

# Drop cities whose crime counts are entirely zero. Columbus's upstream bg_crime extract produced
# all-zero counts (BG geometry present, crime never joined) -> a constant rate that makes
# correlation / ranking metrics undefined. Remove it rather than let it poison the LEVEL metrics.
_degenerate = [c for c, g in pool10.groupby("city") if float(g["wprop_rate"].abs().sum()) == 0]
if _degenerate:
    print("dropping all-zero-crime cities (upstream data gap):", _degenerate)
    pool10 = pool10[~pool10["city"].isin(_degenerate)].copy()
    all_cities = [c for c in all_cities if c not in _degenerate]

print("pool5:", pool5.shape, "| pool10:", pool10.shape, "| wprop cities:", len(all_cities))

  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  kansas_city         473 BGs
  detroit             622 BGs
  pooled             5314 BGs across 5 cities
  dropped 22 zero/NaN-pop BGs -> 5292 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 8 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 5284 remain
  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  sacramento          366 BGs
  san_francisco       679 BGs
  pittsburgh          317 BGs
  kansas_city         473 BGs
  columbus            637 BGs
  jacksonville        550 BGs
  detroit             622 BGs
  pooled             7863 BGs across 10 cities


  dropped 32 zero/NaN-pop BGs -> 7831 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 12 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 7819 remain
dropping all-zero-crime cities (upstream data gap): ['columbus']
pool5: (5284, 108) | pool10: (7182, 108) | wprop cities: 9


### Benchmark & metric glossary

The `existing (04)` row is the deployed **agency-scale** model, scored with the **same helpers** as
the refresh (via `baseline.existing_model_run`) so the comparison is apples-to-apples. Its numbers are
its own block-group fit from `04_bg_comparison`, computed live for **both** targets — wtotal / 5 cities
and wprop / 9 cities.

All metrics are on the absolute `log1p(weighted rate)` scale, pooled across held-out block groups:

- **`r2_oos`** — pooled out-of-sample R² = 1 − SSE/SST. The headline calibration number. The incumbent
  scores **0.16** (total) / **0.06** (property) here — it predicts a rate but a miscalibrated one; the
  refresh predicts the rate directly and should beat it.
- **`mae`** — mean absolute error on the same log-rate scale.
- **`recall@25`** — of a city's truly-dangerous blocks (top decile by observed rate), the share the
  model also ranks in its own within-city top 25% (higher = better).
- **`called_safe`** — share of dangerous blocks buried below the model's within-city median (the costly
  miss; lower = better).

The incumbent row is a single **deployed** number (it has seen every city); every refresh row is honest
**LOCO** (city fully held out) or optimistic **80/20** — so *beating the incumbent under LOCO is the
strong result*.


In [3]:
_POOL = {"wtotal": pool5, "wprop": pool10}
_RUNS = {}


def get_run(category, pset, split, model="ridge"):
    """Return a cached absolute log-rate run for one (predictor set, split, model).

    Runs are memoised in ``_RUNS`` so each configuration is fit once, and fitter stdout is
    suppressed to keep tables clean. ``pset`` is 'ours' | 'ours+approved' |
    'ours+approved+agency'; ``split`` is 'loco' | '80_20'; ``model`` is 'ridge' | 'gbm'.
    """
    key = (category, pset, split, model)
    if key not in _RUNS:
        spec = ExperimentSpec(pset, "absolute", split, model)
        with contextlib.redirect_stdout(io.StringIO()):
            _RUNS[key] = run_spec(spec, _POOL[category], category)
    return _RUNS[key]


def pooled_r2_mae(run, category):
    """Pooled out-of-sample R2 and MAE on the log1p(rate) scale (same recipe as 04_bg_comparison):
    y = log1p(observed rate), yhat = the run's y_pred (already log-rate), over every held-out BG."""
    g = run["scored"]
    y = np.log1p(g[f"{category}_rate"].astype(float))
    yhat = g["y_pred"].astype(float)
    resid = y - yhat
    r2 = 1 - (resid ** 2).sum() / ((y - y.mean()) ** 2).sum()
    return round(float(r2), 3), round(float(resid.abs().mean()), 3)


def incumbent_run(category, cities):
    """Deployed incumbent (agency-scale, 04_bg_comparison) as a refresh-style run dict."""
    with contextlib.redirect_stdout(io.StringIO()):
        return baseline.existing_model_run(category, cities=cities)


def incumbent_row(category, cities):
    """Score the incumbent with the SAME helpers as the refresh -> its reference row.

    ``total_pt_ct`` / ``property_pt_ct`` IS a predicted weighted crime rate, so it carries a real
    (if weak) ``r2_oos`` / ``mae`` -- the as-served block-group fit reported in 04 -- alongside the
    within-city ``recall@25`` / ``called_safe``.
    """
    run = incumbent_run(category, cities)
    r2, mae = pooled_r2_mae(run, category)
    rec = within_city_recall(run, category, nets=(0.25,)).loc["POOLED"]
    return {"r2_oos": r2, "mae": mae,
            "recall@25": rec["recall@top25"], "called_safe": rec["called_safe"]}


# Incumbent reference rows, computed live for both targets (wtotal / 5, wprop / 9 cities).
EXISTING_WTOTAL = incumbent_row("wtotal", full_cities)
EXISTING_WPROP = incumbent_row("wprop", all_cities)


def combined_table(category, sets, model="ridge", existing=None):
    """Calibration + rank scorecard for one model across predictor sets, LOCO and 80/20.

    CALIBRATION columns (block placed at the right rate): r2_oos, mae.
    RANK columns (dangerous-block flagging within a city): recall@25, called_safe.
    An optional ``existing`` dict prepends the deployed incumbent reference row (its own 04 fit).
    """
    rows = []
    if existing is not None:
        rows.append({"predictors": "existing (04)", "split": "deployed", **existing})
    # One block per predictor set; within each, the honest LOCO row then the optimistic 80/20 row.
    for name in sets:
        for split in ["loco", "80_20"]:
            run = get_run(category, name, split, model)
            r2, mae = pooled_r2_mae(run, category)
            rec = within_city_recall(run, category, nets=(0.25,)).loc["POOLED"]
            rows.append({"predictors": name, "split": split.replace("_", "/"),
                         "r2_oos": r2, "mae": mae,
                         "recall@25": rec["recall@top25"], "called_safe": rec["called_safe"]})
    return pd.DataFrame(rows).set_index(["predictors", "split"])


def level_by_city(category, pset, model="ridge", split="loco"):
    """Per-held-out-city decomposition of the pooled r2_oos -- is it uniform or sunk by one city?

    For each held-out city: its observed vs predicted city-MEAN log-rate (`level_err`, the
    calibration miss the agency total is meant to fix) and its own out-of-sample `r2_oos`. A single
    strongly-negative city is what drags the pooled number down (the small-sample risk on wtotal/5).
    """
    run = get_run(category, pset, split, model)
    g = run["scored"].copy()
    g["_obs"] = np.log1p(g[f"{category}_rate"].astype(float))
    rows = []
    # One row per held-out city: its level miss and its own r2_oos.
    for city, d in g.groupby("holdout_city"):
        y, p = d["_obs"], d["y_pred"]
        r2 = 1 - ((y - p) ** 2).sum() / ((y - y.mean()) ** 2).sum()
        rows.append({"city": city, "n": len(d),
                     "obs_mean": round(float(y.mean()), 2), "pred_mean": round(float(p.mean()), 2),
                     "level_err": round(float(p.mean() - y.mean()), 2), "r2_oos": round(float(r2), 3)})
    return pd.DataFrame(rows).set_index("city").sort_values("r2_oos")


### Predictor sets

Exactly what each rung feeds the model. The ladder is additive: `ours+approved` adds the approved,
de-duplicated incumbent features to `ours`; `ours+approved+agency` adds **one** column,
`agency_lag_<target>_log` — last year's agency-wide crime total, identical across every BG in a city
(so it carries the city-level variance; the BG predictors carry the rest).

In [4]:
# Print the resolved predictor columns for each rung (wtotal pool), showing what each rung adds.
_seen = []
for pset in ["ours", "ours+approved", "ours+approved+agency"]:
    cols = resolve_predictors(ExperimentSpec(pset, "absolute", "loco", "ridge"), "wtotal", pool5)
    added = [c for c in cols if c not in _seen]      # only the columns this rung introduces
    print(f"{pset}  (total {len(cols)}, +{len(added)} new):")
    for c in added:
        print("   +", c)
    _seen = cols
    print()

ours  (total 23, +23 new):
   + moved1yr_pct
   + own_pct
   + lap_pct
   + city_centers_dist
   + pop_est_5mile_log
   + pop_ch_1mile
   + vacant_pct_log
   + clip_liens_pct_log
   + clip_foreclosure_pct_log
   + unq_convenience_stores_clips
   + unq_gas_stations_clips
   + unq_liquor_stores_clips
   + vacant_pct_lag6_log
   + clip_liens_pct_lag6_log
   + clip_foreclosure_pct_lag6_log
   + transit_has_transit
   + transit_service_intensity_logc
   + transit_nearest_stop_m_log
   + transit_overnight_stop_share
   + roof_condition_avg
   + roof_debris_pct_avg
   + hardscapes_pct_avg
   + roof_missing_material_pct

ours+approved  (total 28, +5 new):
   + in_household_pct
   + det_pct
   + div_encentral
   + div_midatlantic
   + div_southatlantic

ours+approved+agency  (total 29, +1 new):
   + agency_lag_wtotal_log



### 1 · The incumbent's block-group problem — property *and* total

The deployed agency-scale model fits block-group crime **weakly**, on both crime families (the full
per-city breakdown lives in `04_bg_comparison`; here it is the reference row the refresh must beat):

- **Calibration is poor.** Pooled out-of-sample `r2_oos` is **0.16** (total) / **0.06** (property) on
  the log-rate scale — and several cities go **negative** (the served prediction underperforms a
  per-city constant). **Detroit** is the sharpest case.
- **Within-city ranking is only fair.** `recall@25` ≈ 0.59 (total) / 0.50 (property), `called_safe`
  ≈ 0.16 / 0.20 (dangerous blocks it buries below the city median).

That block-group gap is the whole target of the refresh.


In [5]:
# The incumbent's own block-group fit -- the reference row for every table below (both targets).
print("INCUMBENT (deployed agency-scale) reference -- its own block-group fit, from 04_bg_comparison\n")
print("  wtotal / 5 cities:", EXISTING_WTOTAL)
print("  wprop  / 9 cities:", EXISTING_WPROP)
print("\nWeak calibration (r2_oos 0.16 / 0.06; several cities negative, Detroit worst) and only fair")
print("within-city ranking. See 04_bg_comparison for the per-city detail; the refresh below closes it.")


INCUMBENT (deployed agency-scale) reference -- its own block-group fit, from 04_bg_comparison

  wtotal / 5 cities: {'r2_oos': 0.164, 'mae': 0.787, 'recall@25': np.float64(0.59), 'called_safe': np.float64(0.163)}
  wprop  / 9 cities: {'r2_oos': 0.058, 'mae': 0.712, 'recall@25': np.float64(0.5), 'called_safe': np.float64(0.196)}

Weak calibration (r2_oos 0.16 / 0.06; several cities negative, Detroit worst) and only fair
within-city ranking. See 04_bg_comparison for the per-city detail; the refresh below closes it.


### 2 · The refresh on PROPERTY crime (9 cities)

We lead with **property** because it has the **wider sample** (9 cities vs 5), which makes every
cross-city number less brittle — the clearest way to see that **sample size, not model cleverness, is
the main lever.** Same predictor ladder (`ours → +approved → +agency`), both models, scored on LEVEL
and RANK against the incumbent benchmark from section 1.

Read the two splits as a pair: **80/20** (a city we've already labeled — interpolation) vs **LOCO** (a
brand-new city — extrapolation). Watch the agency lagged total lift cross-city LEVEL, and watch the
80/20→LOCO gap as the price of never having seen a city.


In [6]:
SETS = ["ours", "ours+approved", "ours+approved+agency"]
print(f"BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · RidgeCV")
print(combined_table("wprop", SETS, model="ridge", existing=EXISTING_WPROP).to_string())
print("\n* existing (04) is the deployed model's own block-group fit (its reference row).")
print("  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.")


BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / 9 cities · RidgeCV


                               r2_oos    mae  recall@25  called_safe
predictors           split                                          
existing (04)        deployed   0.058  0.712      0.500        0.196
ours                 loco       0.327  0.595      0.658        0.119
                     80/20      0.385  0.571      0.624        0.128
ours+approved        loco       0.358  0.579      0.651        0.112
                     80/20      0.417  0.557      0.638        0.142
ours+approved+agency loco       0.380  0.566      0.652        0.113
                     80/20      0.422  0.554      0.617        0.142

* existing (04) is the deployed model's own block-group fit (its reference row).
  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.


In [7]:
print(f"BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · LightGBM")
print(combined_table("wprop", SETS, model="gbm", existing=EXISTING_WPROP).to_string())
print("\n* same predictor ladder; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.")


BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / 9 cities · LightGBM


                               r2_oos    mae  recall@25  called_safe
predictors           split                                          
existing (04)        deployed   0.058  0.712      0.500        0.196
ours                 loco       0.270  0.633      0.630        0.117
                     80/20      0.460  0.537      0.645        0.135
ours+approved        loco       0.360  0.592      0.630        0.114
                     80/20      0.483  0.527      0.660        0.099
ours+approved+agency loco       0.359  0.594      0.642        0.123
                     80/20      0.490  0.521      0.674        0.106

* same predictor ladder; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.


**Read — property refresh (9 cities).**

- **The refresh calibrates block-group crime the incumbent can't.** RidgeCV LOCO `r2_oos` climbs
  **0.33 → 0.36 → 0.38** across the ladder — against the incumbent's **0.06**. We predict a genuine
  per-BG rate; the incumbent's served rate barely beats a constant.
- **We beat it on rank too, under honest LOCO.** `recall@25` ≈ 0.65 vs **0.50**, `called_safe`
  ≈ 0.11–0.12 vs **0.20** (roughly half the costly misses).
- **The agency lagged total is a decisive calibration predictor.** It lifts Ridge LOCO `r2_oos`
  0.33 → 0.38 (80/20 0.39 → 0.42) — it carries the city-level rate every BG in a city shares, so the
  BG predictors are freed to carry the rest.
- **Sample size is the main lever, not model cleverness.** Every 80/20 `r2_oos` beats its LOCO twin
  (e.g. `ours` Ridge 0.39 vs 0.33) — the gap is purely "have we labeled this city," and it is the
  biggest swing on the page.
- **Ridge vs GBM: Ridge extrapolates more stably.** At the top rung Ridge LOCO `r2_oos` **0.380** > GBM
  **0.359**, and GBM's 80/20→LOCO drop is far larger (0.490 → 0.359 = 0.13) than Ridge's
  (0.422 → 0.380 = 0.04) — GBM overfits the cities it has seen. On rank the two are a near-tie.


### 3 · The same refresh on TOTAL crime (5 cities)

Weighted-**total** covers only the 5 full-coverage cities, so its cross-city calibration rests on 5
city levels — powerful but fragile (the per-city breakdown below opens it up). Same ladder, both
models, against the wtotal incumbent reference row.


In [8]:
print("BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV")
print(combined_table("wtotal", SETS, model="ridge", existing=EXISTING_WTOTAL).to_string())
print("\n* existing (04) is the deployed model's own block-group fit. Refresh rows are honest LOCO and 80/20.")


BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV


                               r2_oos    mae  recall@25  called_safe
predictors           split                                          
existing (04)        deployed   0.164  0.787      0.590        0.163
ours                 loco       0.245  0.722      0.728        0.077
                     80/20      0.502  0.599      0.686        0.049
ours+approved        loco       0.232  0.718      0.732        0.079
                     80/20      0.514  0.594      0.696        0.069
ours+approved+agency loco       0.194  0.751      0.716        0.085
                     80/20      0.515  0.593      0.696        0.069

* existing (04) is the deployed model's own block-group fit. Refresh rows are honest LOCO and 80/20.


In [9]:
print("BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 cities · LightGBM")
print(combined_table("wtotal", SETS, model="gbm", existing=EXISTING_WTOTAL).to_string())
print("\n* same predictor ladder as the RidgeCV table; compare LOCO rows head-to-head.")


BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 cities · LightGBM


                               r2_oos    mae  recall@25  called_safe
predictors           split                                          
existing (04)        deployed   0.164  0.787      0.590        0.163
ours                 loco       0.353  0.696      0.694        0.109
                     80/20      0.516  0.604      0.657        0.039
ours+approved        loco       0.379  0.679      0.718        0.095
                     80/20      0.521  0.601      0.676        0.049
ours+approved+agency loco       0.375  0.686      0.710        0.101
                     80/20      0.529  0.596      0.657        0.069

* same predictor ladder as the RidgeCV table; compare LOCO rows head-to-head.


**Read — total refresh (5 cities).**

- **Same rank win over the incumbent, if anything larger.** LOCO `recall@25` ≈ 0.71–0.73 vs **0.59**,
  `called_safe` ≈ 0.08–0.10 vs **0.16** (roughly half the costly misses). This is the Detroit-style
  within-city gap being closed.
- **Calibration beats the incumbent but is fragile on 5 cities.** GBM LOCO `r2_oos` ≈ 0.35–0.38 (vs
  incumbent **0.16**) and holds across the ladder. RidgeCV LOCO `r2_oos` instead *falls*
  0.245 → 0.232 → **0.194** as the agency total is added — the opposite of property — because on only 5
  cities that cross-city signal is brittle (the per-city breakdown shows one city, Atlanta, causes it).
  Both 80/20 rows sit at ~0.50–0.53.
- **The Ridge-vs-GBM verdict flips vs property** (GBM wins calibration on wtotal/5, Ridge on wprop/9) —
  a small-sample artifact, not a real skill gap. Judge extrapolation on the 9-city property evidence.
- **Both models beat the incumbent on rank under LOCO on both targets** — the robust, repeatable result.


### 4 · Per-city breakdown — is the pooled `r2_oos` uniform or carried by 1–2 cities?

Pooled `r2_oos` is a single fit over every held-out block group, so one badly-calibrated city can
dominate it — especially on wtotal's 5. `level_by_city` opens it up per held-out city:
- **`level_err`** = predicted mean − observed mean log-rate (LOCO). Near 0 = the city is placed at the
  right height; a large value is the calibration miss the agency total is meant to fix.
- **`r2_oos`** = that city's own out-of-sample fit. A single strongly-negative city is what drags the
  pooled number down.


In [10]:
# Decompose each config's pooled r2_oos per held-out city on the top rung (ours+approved+agency).
CFG = "ours+approved+agency"
for cat in ["wtotal", "wprop"]:
    # One target at a time; Ridge then GBM so the level_err / r2_oos columns sit side by side.
    for model in ["ridge", "gbm"]:
        print(f"== {cat} · {CFG} · {model} · LOCO (worst-calibrated city first) ==")
        print(level_by_city(cat, CFG, model=model).to_string())
        print()


== wtotal · ours+approved+agency · ridge · LOCO (worst-calibrated city first) ==
                n  obs_mean  pred_mean  level_err  r2_oos
city                                                     
atlanta       369      3.41       5.28       1.87  -1.990
detroit       600      4.20       4.55       0.35  -0.020
houston      1536      3.45       3.59       0.15   0.336
chicago      2064      3.44       3.10      -0.34   0.367
kansas_city   445      3.74       3.57      -0.18   0.559

== wtotal · ours+approved+agency · gbm · LOCO (worst-calibrated city first) ==
                n  obs_mean  pred_mean  level_err  r2_oos
city                                                     
detroit       600      4.20       4.49       0.29   0.088
atlanta       369      3.41       3.99       0.57   0.202
houston      1536      3.45       3.28      -0.16   0.288
chicago      2064      3.44       3.59       0.15   0.421
kansas_city   445      3.74       3.44      -0.30   0.489

== wprop · ours+approved+a

**Read — per-city breakdown (why the pooled `r2_oos` moves).** The pooled number is one fit over all
held-out blocks, so one miscalibrated city dominates it.

- **wtotal / 5 — Ridge is sunk by a single city.** Atlanta `level_err` **+1.87** (predicts mean 5.28
  vs observed 3.41) gives it a per-city `r2_oos` of **−1.99**, dragging Ridge's pooled LOCO `r2_oos` to
  **0.194** while the other four cities sit at 0.34–0.56. GBM has no blow-up (Atlanta `level_err` +0.57,
  `r2_oos` 0.20) → pooled **0.375**. This lone outlier — not a general skill gap — is the whole wtotal
  calibration story. Do **not** tune toward cross-city `r2_oos` on n = 5.
- **wprop / 9 — with the agency total, Ridge is well-calibrated everywhere.** Every city's `level_err`
  is within ±0.17 and every per-city `r2_oos` is positive (0.03–0.47) → pooled **0.380**. GBM instead
  over-predicts Atlanta (**+0.54** → `r2_oos` 0.09) and San Francisco (+0.28 → 0.11) → pooled **0.359**.
  The linear model's even calibration is what wins on the wider target.
- **Rank stays healthy for both regardless of the calibration wobble** — judge extrapolation on the
  9-city property evidence, not wtotal/5.


### 5 · Hyperparameter tuning — does Optuna move LightGBM?

The GBM above uses fixed defaults. Here we let **Optuna** search LightGBM's hyperparameters, tuning
**directly on the LOCO `r2_oos`** — the leave-one-city-out number that reflects the national rollout,
the metric we actually want to maximise.

- **Objective = LOCO `r2_oos`.** We optimise the deployment metric itself rather than an 80/20 proxy,
  so any lift lands on the axis that matters for unseen cities.
- **Honesty caveat.** With only 9 cities, the reported best LOCO `r2_oos` is the *max over 40 trials*
  on the same 9-fold structure, so it carries some selection optimism — read the magnitude as an
  upper bound, and check the per-city breakdown that the lift isn't carried by a single city. The
  **80/20** column below is the untouched cross-check.


In [11]:
# Tune LightGBM on the PROPERTY target (9 cities), the wider/less-brittle sample.
TUNE_CAT, TUNE_PSET = "wprop", "ours+approved+agency"


def _gbm_pooled(params, split, field):
    """Fit tuned-GBM through the same run_spec pipeline; return one POOLED metric."""
    spec = ExperimentSpec(TUNE_PSET, "absolute", split, "gbm", gbm_params=params)
    with contextlib.redirect_stdout(io.StringIO()):
        run = run_spec(spec, pool10, TUNE_CAT)
    return float(loco_metrics(run).loc["POOLED"][field])


def _objective(trial):
    """Maximise LOCO r2_oos directly -- the leave-one-city-out (extrapolation) deployment metric."""
    params = dict(
        n_estimators=trial.suggest_int("n_estimators", 200, 1200, step=100),
        learning_rate=trial.suggest_float("learning_rate", 0.01, 0.1, log=True),
        num_leaves=trial.suggest_int("num_leaves", 8, 127),
        min_child_samples=trial.suggest_int("min_child_samples", 10, 150),
        subsample=trial.suggest_float("subsample", 0.6, 1.0),
        colsample_bytree=trial.suggest_float("colsample_bytree", 0.6, 1.0),
        reg_alpha=trial.suggest_float("reg_alpha", 0.0, 5.0),
        reg_lambda=trial.suggest_float("reg_lambda", 0.0, 5.0),
    )
    return _gbm_pooled(params, "loco", "r2_oos")


study = optuna.create_study(direction="maximize", sampler=optuna.samplers.TPESampler(seed=42))
study.optimize(_objective, n_trials=40, show_progress_bar=False)
best = study.best_params
print("best LOCO r2_oos:", round(study.best_value, 4))
print("best params:", {k: (round(v, 4) if isinstance(v, float) else v) for k, v in best.items()})


def _scorecard(params, tag):
    """Print LOCO + 80/20 calibration/rank for a param set through the same pipeline."""
    for split in ["loco", "80_20"]:
        spec = ExperimentSpec(TUNE_PSET, "absolute", split, "gbm", gbm_params=params)
        with contextlib.redirect_stdout(io.StringIO()):
            run = run_spec(spec, pool10, TUNE_CAT)
        m = loco_metrics(run).loc["POOLED"]
        rec = within_city_recall(run, TUNE_CAT, nets=(0.25,)).loc["POOLED"]
        print(f"  {tag:8s} {split:5s} | r2_oos={m['r2_oos']:.3f} mae={m['mae']:.3f} "
              f"recall@25={rec['recall@top25']:.3f} called_safe={rec['called_safe']:.3f}")


print("\nwprop · default vs Optuna-tuned LightGBM (objective = LOCO r2_oos; 80/20 is the cross-check):")
_scorecard(None, "default")
_scorecard(best, "tuned")

best LOCO r2_oos: 0.378
best params: {'n_estimators': 700, 'learning_rate': 0.0131, 'num_leaves': 122, 'min_child_samples': 105, 'subsample': 0.9981, 'colsample_bytree': 0.9378, 'reg_alpha': 4.1183, 'reg_lambda': 4.4776}

wprop · default vs Optuna-tuned LightGBM (objective = LOCO r2_oos; 80/20 is the cross-check):


  default  loco  | r2_oos=0.359 mae=0.590 recall@25=0.642 called_safe=0.123


  default  80_20 | r2_oos=0.490 mae=0.520 recall@25=0.674 called_safe=0.106


  tuned    loco  | r2_oos=0.378 mae=0.580 recall@25=0.654 called_safe=0.117


  tuned    80_20 | r2_oos=0.476 mae=0.520 recall@25=0.674 called_safe=0.106


**Read — tuning.** Optuna optimises the LOCO `r2_oos` directly and buys a **modest, honest lift**:

- LOCO `r2_oos` **0.359 → 0.378** (+0.019), `mae` 0.590 → 0.580, `recall@25` 0.642 → 0.654,
  `called_safe` 0.123 → 0.117 — every metric nudges the right way. It is essentially the same ceiling
  (~0.38) the 80/20-tuned model reached, so tuning *toward* LOCO doesn't unlock more than the proxy did.
- **It's genuine extrapolation, not interpolation.** The tuned model's untouched **80/20** `r2_oos`
  actually slips (0.490 → 0.476) — Optuna traded a little in-sample fit for the LOCO gain, exactly what
  we want.
- **Verdict: a real but second-order lever.** Best params ask for heavy regularization (slow
  `learning_rate` ≈ 0.013, strong `min_child_samples` 105 / L1–L2 ≈ 4 — Optuna wants *less*
  overfitting). Worth keeping, but small next to the two first-order levers: adding the **agency
  predictor** and adding **cities**.


### Next steps

1. **More cities — the highest-leverage move.** Every number here rests on 5–9 cities, and LOCO is
   harsh precisely because each held-out city is structurally unique. Adding POC cities shrinks the
   80/20→LOCO gap and de-fragilizes the cross-city `r2_oos` far more than any modelling change.
2. **Tuning (section 5).** Optuna buys a modest, honest lift on LOCO calibration — worth keeping, but
   it is a second-order lever behind sample size and features. Do **not** tune toward the cross-city
   `r2_oos` on this few cities.
3. **Predictors — target *within-city* signal.** The agency lag already carries the between-city level;
   new features should explain block-to-block variation (the incumbent's weak spot and Detroit's):
   finer POI mix, local built-environment / imagery, temporal crime lags where available, and
   mobility / foot-traffic. Brainstorm and screen these the same way — LOCO first, RANK metrics first.
